### ISSUE TITLE
Feature request:  Hamming distance for non-binary vectors

### ISSUE THREAD

**Issue by lesshaste** [nmslib/nmslib#340]
I hope it isn't inappropriate to make feature requests, but it would be great if nmslib supported Hamming distance for alphabets of size greater than 2.

**Comment by searchivarius** [comment]
@lesshaste do you mean strings or integers?

**Comment by lesshaste** [comment]
I don't really make a distinction but let me be more precise. I will write HD for Hamming distance.

HD('0215', 0211') = 1

HD([0,2,1,5], [0,2,1,1]) = 1

So the difference is just what is convenient for you as coders.

**Comment by VladVin** [comment]
I also find it useful to support other alphabets except binary. Look, for example, at [ORB descriptors](https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_feature2d/py_orb/py_orb.html#orb-in-opencv) in OpenCV, specifically parameter `WTA_K`. When it's 3 or 4, the hamming distance is calculated using 2 bits, and OpenCV supports that under `normType=cv2.NORM_HAMMING2`

**Comment by searchivarius** [comment]
@VladVin I don't think it's a commonly used feature, but it's not so difficult to implement anyways. Pull requests are welcome.

**Comment by lesshaste** [comment]
I do feel it would be a great addition.  I can think of quite a lot of uses myself.

**Comment by searchivarius** [comment]
Right, it might be a bit tricky, b/c we would have to have different implementations optimized for different alphabet sizes. One might even want to consider the vertical layout that we used for our SIMD compression. Again, how far down the rabbit hole should one go?
